components/automate-cli/pkg/verifyserver/services/fqdnservice/fqdnservice.go

```diff
@@ -0,0 +1,240 @@
+				fqdnResultChan <- "got error"
+				return
+			}
+			nodeIP := res.Header.Get("x-server-ip")
+			fqdnResultChan <- nodeIP
+		}(fqdnResultChan)
+	}
+
+	for i := 0; i < 50; i++ {
+		chanResult := <-fqdnResultChan
+		if chanResult == "got error" {
+			continue
+		}
+		delete(setNodes, chanResult)
+		//if setNodes becomes empty, that means we are able to reach all the nodes given in the request body.
+		if len(setNodes) == 0 {
+			fq.log.Debug("All nodes are reachable.")
+			return createCheck(constants.NODE_TITLE, true, constants.NODE_SUCCESS_MESSAGE, "", "")
+		}
+	}
+
+	errorMessage := createErrorMessage(setNodes, reqNodes, isAfterDeployment)
+	return createCheck(constants.NODE_TITLE, false, "", errorMessage, constants.NODE_RESOLUTION_MESSAGE)
+}
+
+// validateCertificate will check that if our root certificate is valid or not.
+func (fq *FqdnService) validateCertificate(rootCert string) models.Checks {
+	fq.log.Debug("Validating Certificate...")
+	// Parse the PEM-encoded SSL certificate
+	block, _ := pem.Decode([]byte(rootCert))
+	if block == nil || block.Type != "CERTIFICATE" {
+		fq.log.Error("Failed to decode certificate PEM")
+		return createCheck(constants.CERTIFICATE_TITLE, false, "", constants.CERTIFICATE_ERROR_MESSAGE, constants.CERTIFICATE_RESOLUTION_MESSAGE)
+	}
+
+	certificate, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		fq.log.Error(err.Error())
+		return createCheck(constants.CERTIFICATE_TITLE, false, "", constants.CERTIFICATE_ERROR_MESSAGE, constants.CERTIFICATE_RESOLUTION_MESSAGE)
+	}
+
+	currentTime := time.Now()
+	if currentTime.After(certificate.NotAfter) || !certificate.IsCA || len(certificate.DNSNames) == 0 {
+		fq.log.Debug("Certificate is Invalid.")
+		fq.log.Debugf("Expiry: %v, CA: %v, SAN: %v", certificate.NotAfter, certificate.IsCA, certificate.DNSNames)
+		return createCheck(constants.CERTIFICATE_TITLE, false, "", constants.CERTIFICATE_ERROR_MESSAGE, constants.CERTIFICATE_RESOLUTION_MESSAGE)
+	}
+
+	fq.log.Debug("Certificate is valid")
+	return createCheck(constants.CERTIFICATE_TITLE, true, constants.CERTIFICATE_SUCCESS_MESSAGE, "", "")
+}
+
+func createCheck(title string, passed bool, successMsg, errorMsg, resolutionMsg string) models.Checks {
+	return models.Checks{
+		Title:         title,
+		Passed:        passed,
+		SuccessMsg:    successMsg,
+		ErrorMsg:      errorMsg,
+		ResolutionMsg: resolutionMsg,
+	}
+}
+
+func (fq *FqdnService) CheckFqdnReachability(req models.FqdnRequest, port string) models.FqdnResponse {
+	var response = models.FqdnResponse{}
+
+	check := fq.fqdnReachable(req.Fqdn, req.RootCert, req.NodeType, req.IsAfterDeployment, port)
+	response.Checks = append(response.Checks, check)
+
+	check = fq.nodeReachable(req.Fqdn, req.RootCert, req.Nodes, req.IsAfterDeployment, port)
+	response.Checks = append(response.Checks, check)
+
+	if !req.IsAfterDeployment {
+		check = fq.validateCertificate(req.RootCert)
+		response.Checks = append(response.Checks, check)
```

Comment:
should we validate certificates before checking the reachability ? What if the certificates are invalid hen why are checking the nodes and fqdn reahbaility?

Reply:
added conditions